Unique instruction ids

alan-baker opened the pull request:
In preparation for #846, this request adds an unique id to all instructions. This id is issued by the IRContext. This means instructions are effectively owned by a context. For convenience, Instruction now caches a pointer to the IRContext.

* Instruction constructors now take a context where possible
  * could not remove default or copy constructor yet, but they should **NOT** be used explicitly
* Removed instances of instruction/block/function copies or moves and replaced them with Clones.
* Removed copy constructors for blocks and functions. 
* Added an internal ability for the linker to verify unique-ness of ids in the merged module (realistically the only place dealing with multiple contexts).

s-perron commented on source/opt/decoration_manager.cpp:
```diff
@@ -261,7 +261,7 @@ void DecorationManager::CloneDecorations(
         // simply clone decoration and change |target-id| to |to|
-        std::unique_ptr<ir::Instruction> new_inst(inst->Clone());
+        std::unique_ptr<ir::Instruction> new_inst(inst->Clone(module_->context()));
         new_inst->SetInOperand(0, {to});
```
Not for this PR, but I forgot to move this function to the IRContext.

s-perron commented on source/opt/instruction.h:
```diff
@@ -83,29 +84,30 @@ class Instruction : public utils::IntrusiveNodeBase<Instruction> {
   using const_iterator = std::vector<Operand>::const_iterator;
 
-  // Creates a default OpNop instruction.
+  // This exists solely for containers that can't do without. Should be removed.
```
Should still keep the comment about what it does, even if we want to eventually remove it.

s-perron commented on source/opt/instruction.h:
```diff
@@ -241,15 +246,22 @@ class Instruction : public utils::IntrusiveNodeBase<Instruction> {
+  inline bool operator<(const Instruction&) const;
+
  private:
+
```
No blank line here.

dnovillo commented on include/spirv-tools/linker.hpp:
```diff
@@ -41,8 +43,16 @@ class LinkerOptions {
 
+  // Returns whether to verify the ids in the merged module
+  bool GetVerifyIds() const { return verifyIds_; }
+  // Sets whether to verify the ids in the merged module
```
Blank line above.

dnovillo commented on source/link/linker.cpp:
```diff
@@ -186,7 +192,8 @@ spv_result_t Linker::Link(const uint32_t* const* binaries,
 
-  std::vector<std::unique_ptr<Module>> modules;
+  std::vector<std::unique_ptr<IRContext>> contexts;
+  std::vector<Module*> modules;
```
Curious why did the vector of Modules change from a vector of unique_ptr to a vector of raw ptr?

alan-baker replied:
Convenience. The modules are owned by the vector<unique_ptr<IRContext>>, but most of the code referred to the module vector. So to avoid further rewrites I also tracked raw pointers to the modules to avoid larger changes in the linker.

dnovillo commented on source/opt/build_module.h:
```diff
@@ -18,23 +18,25 @@
-std::unique_ptr<ir::Module> BuildModule(spv_target_env env,
+// Builds an ir::Module returns the owning ir::IRContext from the given SPIR-V
+// |binary|. |size| specifies number of words in |binary|. The |binary| will be
+// decoded according to the given target |env|. Returns nullptr if erors occur
```
s/erors/errors/

dnovillo commented on source/opt/build_module.h:
```diff
@@ -18,23 +18,25 @@
-std::unique_ptr<ir::Module> BuildModule(
+// Builds an ir::Module and returns the owning ir::IRContext from the given
+// SPIR-V assembly |text|.  The |text| will be encoded according to the given
+// target |env|. Returns nullptr if erors occur and sends the errors to
```
s/erors/errors/  (the misspelling was there before, but since you're here... ;)

dnovillo commented on source/opt/ir_context.h:
```diff
@@ -239,6 +263,14 @@ class IRContext {
+  inline uint32_t TakeNextUniqueId() {
+    assert(unique_id_ != std::numeric_limits<uint32_t>::max());
+
+    // Skip zero
```
Terminate all comments with '.' (yeah, we're not consistent :)

dnovillo commented on source/opt/module.h:
```diff
@@ -223,11 +225,17 @@ class Module {
 
+  // Sets the associated context for this module
+  void SetContext(IRContext* c) { context_ = c; }
+  // Gets the associated context for this module
```
Blank line above, please.

dneto0 commented on source/opt/ir_context.h:
```diff
@@ -53,11 +54,34 @@ class IRContext {
   friend inline Analysis operator<<(Analysis a, int shift);
   friend inline Analysis& operator<<=(Analysis& a, int shift);
 
+  // Trivial constructor for testing purposes
```
What goes wrong if someone uses this by accident?  Seems like we should more strenuously steer people away from this.

If you truly need something for testing purposes, then I recommend making a static method to make it, eg..

static IRContext* MakeIRContextForTest() ... and then hide the details


...
Then I read the tests and made a suggestion there to use a helper function to make the dummy IRContexts.

dneto0 commented on source/opt/ir_context.h:
```diff
@@ -239,6 +263,14 @@ class IRContext {
   // Kill all name and decorate ops targeting the result id of |inst|.
   void KillNamesAndDecorates(ir::Instruction* inst);
 
+  // Return the next unique id for use by an instruction
```
While we're nitpicky, prefer "Returns" over "Return".
See style guide:  https://google.github.io/styleguide/cppguide.html#Function_Comments 
In particular:  "These comments should be descriptive ("Opens the file") rather than imperative ("Open the file"); the comment describes the function, it does not tell the function what to do"

Get used to writing comments in that style.  Yes, the codebase is inconsistent on this, but I'm holding Googlers to the Google style guide for this rule.

dneto0 commented on source/opt/ir_context.h:
```diff
@@ -264,6 +296,7 @@ class IRContext {
     valid_analyses_ = valid_analyses_ | kAnalysisDecorations;
   }
 
+  uint32_t unique_id_;
```
Needs a comment. to describe this, especially how it starts at 0 but is not the next unique ID to be used.  It's one *less* than the next unique id to be used.

dneto0 commented on source/opt/ir_loader.cpp:
```diff
@@ -20,22 +20,22 @@
 namespace ir {
 
-IrLoader::IrLoader(const MessageConsumer& consumer, Module* module)
+IrLoader::IrLoader(const MessageConsumer& consumer, Module* m)
```
Are you doing this to avoid shadowing?  Probably, and that's ok. :-)

alan-baker replied:
Yes this is because of shadowing.

dneto0 commented on test/opt/instruction_test.cpp:
```diff
@@ -168,7 +175,8 @@ TEST(InstructionTest, OperandConstIterators) {
 
 TEST(InstructionTest, OperandIterators) {
-  Instruction inst(kSampleParsedInstruction);
+  IRContext context;
```
Instead of having a default constructor that nobody else in the world uses, I recommend making a helper function in this module to make a dummy, but with few words.

e.g. 
static IRContext DummyContext() { return IRContext(Module(), ...); }

Then in the functions:
IRContext& context = DummyContext();

And then let lifetime extension carry it through.

alan-baker commented:
I removed the default IRContext and while I was at it I reworked some large portions of the DefUseManager tests in anticipation of the next round of changes. Previously I was creating dummy contexts, but this would have needed to be rewritten when the storage of uses changes anyways. Removing the temporary contexts required changing the way several of the tests functioned. I removed some that no longer made sense and replaced others that were awkward to instantiate via the parameterizations. Unfortunately, this has increased the size of the diff considerably.

dneto0 commented:
Rebased, squashed, and pushed into master as a771713e4
